### PR TITLE
Better preservation of names

### PR DESCRIPTION
**Pull request by inducer**
This contains some work towards better retention of names as code makes it through the lazy infrastructure.

Best read commit-by-commit (each commit is tiny).

### CODE REVIEW

File: arraycontext/impl/pytato/__init__.py [hunk 1]
```diff
@@ -47,14 +47,39 @@
...
 import abc
 
 if TYPE_CHECKING:
     import pytato
 
 
+def _preprocess_array_tags(tags: ToTagSetConvertible) -> FrozenSet[Tag]:
+    tags = normalize_tags(tags)
+
+    name_hints = [tag for tag in tags if isinstance(tag, NameHint)]
```
Comment (kaushikcfd) [inline arraycontext/impl/pytato/__init__.py:61]: Add vim fold-markers for preprocessing the `NameHint` tags? I believe there would be more such mappings in the near future.

File: arraycontext/impl/pytato/__init__.py [hunk 2]
```diff
@@ -47,14 +47,39 @@
...
+    if name_hints:
+        name_hint, = name_hints
```
Comment (kaushikcfd) [inline arraycontext/impl/pytato/__init__.py:63]: Unhelpful error message if the user passes in >1 `NameHint` tags.

**Review by kaushikcfd — APPROVED** [review]
LGTM, thanks this is helpful in making the generated code more readable while being fully `arraycontext` compatible.

File: arraycontext/impl/pytato/__init__.py (hunk 2 above)
Comment (inducer) [reply arraycontext/impl/pytato/__init__.py:63]: Unlikely to happen given that it's a `UniqueTag`.

File: arraycontext/impl/pytato/__init__.py (hunk 1 above)
Comment (inducer) [reply arraycontext/impl/pytato/__init__.py:63]: https://github.com/inducer/arraycontext/compare/98a3f9116bdc80eccf623edd42ff35ed95b024fa..327103662e4c0c220ae35d12c531f0a5f1254541

File: arraycontext/impl/pytato/__init__.py [hunk 3]
```diff
@@ -47,14 +47,39 @@
...
+            prefix_named, = prefix_nameds
+            from warnings import warn
+            warn("When converting a "
+                    f"arraycontext.metadata.NameHint('{name_hint.name}') "
+                    "to pytato.tags.PrefixNamed, "
+                    f"PrefixNamed('{prefix_named.prefix}') "
+                    "was already present.")
+
+        # type-ignore because mypy does not understand the @tag_dataclass decorator
+        tags = tags | frozenset({PrefixNamed(name_hint.name)})  # type: ignore[call-arg]  # noqa: E501
```
[the remark this answers is not in the document]
Comment (inducer) [reply arraycontext/impl/pytato/__init__.py]: Done: https://github.com/inducer/arraycontext/compare/98a3f9116bdc80eccf623edd42ff35ed95b024fa..327103662e4c0c220ae35d12c531f0a5f1254541

File: doc/other.rst [hunk 4]
```diff
@@ -1,6 +1,13 @@
 Other functionality
 ===================
 
+.. _metadata:
+
+Metadata ("tags" for Arrays and Array Axes
```
[the remark this answers is not in the document]
Comment (inducer) [reply doc/other.rst]: Done: https://github.com/inducer/arraycontext/compare/98a3f9116bdc80eccf623edd42ff35ed95b024fa..327103662e4c0c220ae35d12c531f0a5f1254541

**Comment by inducer** [comment]
Thanks for taking a look!

File: arraycontext/impl/pytato/__init__.py (hunk 2 above)
Comment (kaushikcfd) [reply arraycontext/impl/pytato/__init__.py:65]: Makes sense!

**Pull request merged**